File: leetcode/02311_longest_binary_subsequence_less_than_or_equal_to_k.py
class Solution:
    def longestSubsequence(self, s: str, k: int) -> int:
        ans, num, left, zeroes = 0, 0, -1, 0
        for i, c in enumerate(s):
            num = (num << 1) | (c == '1')
            while num > k:
                left += 1
                if s[left] == '1':
                    num -= 1 << (i - left)
                else:
                    zeroes += 1
            ans = max(ans, zeroes + i - left)
        return ans

File: leetcode/02311_longest_binary_subsequence_less_than_or_equal_to_k.cpp
class Solution {
public:
    int longestSubsequence(string s, int k) {
        int ans = 0, num = 0, zeroes = 0;
        const char *left = s.c_str() - 1;
        for (const char *c = s.c_str(); *c; ++c) {
            for (num = (num << 1) | (*c == '1'); num > k; )
                if (*++left == '1') num ^= 1 << (c - left);
                else ++zeroes;
            int tmp = zeroes + c - left;
            ans = tmp > ans ? tmp : ans;
        }
        return ans;
    }
};

